### PR TITLE
Rearrange vector ops to easily support future changes

### PR DESCRIPTION
**Pull request by Dekken**

### CODE REVIEW

File: lib/include/tick/array/vector_operations.h [hunk 1]
```diff
@@ -1,112 +1,41 @@
...
 #include <numeric>
-
 #include <algorithm>
-#include <atomic>
 #include <type_traits>
 
 #include "promote.h"
 #include "tick/base/defs.h"
 
+#if defined(TICK_MKL_AVAILABLE) || defined(TICK_USE_MKL)
```
Comment (Mbompr) [inline lib/include/tick/array/vector_operations.h]: Why not using only `TICK_USE_MKL` ?

File: lib/include/tick/array/vector_operations.h [hunk 2]
```diff
@@ -1,112 +1,41 @@
...
-};
-
-}  // namespace detail
-}  // namespace tick
-
-#if !defined(TICK_CBLAS_AVAILABLE)
+using vector_operations = detail::vector_operations_mkl<T>;
+}
 
+#elif defined(TICK_CBLAS_AVAILABLE) || defined(TICK_USE_CBLAS)
```
Comment (Mbompr) [inline lib/include/tick/array/vector_operations.h]: Why not only `TICK_USE_CBLAS `?

File: lib/include/tick/array/vector/cblas.h [hunk 3]
```diff
@@ -1,6 +1,24 @@
 
 
-#if defined(TICK_CBLAS_AVAILABLE)
+#ifndef LIB_INCLUDE_TICK_ARRAY_VECTOR_CBLAS_H_
+#define LIB_INCLUDE_TICK_ARRAY_VECTOR_CBLAS_H_
+
```
Comment (Mbompr) [inline lib/include/tick/array/vector/cblas.h]: I would add
`#if defined(TICK_USE_CBLAS)`
Even if it is only included if `TICK_USE_CBLAS` is defined.

File: lib/include/tick/array/vector/cblas.h [hunk 4]
```diff
@@ -1,6 +1,24 @@
```
Comment (Mbompr) [inline lib/include/tick/array/vector/cblas.h]: I like the way you've called mkl ops_mkl, why not ops_blas and ops_unoptimized ?

File: lib/include/tick/array/vector/ops_mkl.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (Mbompr) [inline lib/include/tick/array/vector/ops_mkl.h]: Why do we implement the functions in cblas.h but not in ops_mkl.h? 
Should we have a third file that would call the functions `cblas_*` and include it in both files?
But maybe there is a reason for this dissymmetry.

File: lib/include/tick/array/vector/unoptimized_impl.h [hunk 6]
```diff
@@ -1,5 +1,8 @@
```
Comment (Mbompr) [inline lib/include/tick/array/vector/unoptimized_impl.h]: why not .hpp anymore?

File: lib/include/tick/array/vector/unoptimized_impl.h [hunk 7]
```diff
@@ -1,5 +1,8 @@
 
+#ifndef LIB_INCLUDE_TICK_ARRAY_VECTOR_UNOPTIMIZED_IMPL_H_
+#define LIB_INCLUDE_TICK_ARRAY_VECTOR_UNOPTIMIZED_IMPL_H_
 
+#include <thread>
```
Comment (Mbompr) [inline lib/include/tick/array/vector/unoptimized_impl.h]: can be removed

File: lib/include/tick/array/vector/unoptimized_impl.h (hunk 6 above)
Comment (Dekken) [reply lib/include/tick/array/vector/unoptimized_impl.h]: whoops

Comment (Dekken) [reply lib/include/tick/array/vector/unoptimized_impl.h]: nothing is .hpp - so consistency

File: lib/include/tick/array/vector/ops_mkl.h (hunk 5 above)
Comment (Dekken) [reply lib/include/tick/array/vector/ops_mkl.h]: I think I have a simpler idea

File: lib/CPPLINT.cfg [hunk 8]
```diff
@@ -1,3 +1,3 @@
 set noparent
 linelength=120
-filter=-readability/todo,-legal/copyright,-build/include,-runtime/int,-runtime/references,-whitespace/line_length,-build/c++11
+filter=-readability/todo,-legal/copyright,-build/include,-runtime/int,-runtime/references,-whitespace/line_length,-build/c++11,-readability/casting
```
[the remark this answers is not in the document]
Comment (Dekken) [reply lib/CPPLINT.cfg]: removed the change from CPPLINT.cfg and did

https://github.com/X-DataInitiative/tick/pull/295/files#diff-fa496b9c04c596f1b057030571ee1dacR9 

instead - was complaining about casting

**Pull request merged**